### ISSUE TITLE
Support brick-2 and vty-6.

### ISSUE THREAD

**Issue by amano-kenji** [simonmichael/hledger#2112]
brick-2 depends on vty-6 which introduces breaking changes.

From vty-6 onward, you have to depend on vty-crossplatform, vty-windows, or vty-unix for `mkVty` function.

If you don't use `mkVty` function, then you don't need to depend on one of platform-specific vty packages.

**Comment by ShrykeWindgrace** [comment]
I am working on this. In theory, this requires very minimal change in hledger-ui:
- fix some imports (a couple of lines)
- fix build conditions (delete a couple of lines)
- add lower bounds for brick(/vty/vty-crossplatform, because stack snapshots do not have new versions yet (also a very local change).
- add `notepad` as default editor on windows; git for windows does that, for example (diff is a couple of lines)
- change CI scripts to release windows executable for `hledger-ui`. I haven't looked into that yet; I'll need some help with that

There is, however, a bug:
- run `hledger-ui`
- exit `hledger-ui` (`q`)
- terminal no longer responds to mouse ⚠️ !
I'll see whether the bug is in `hledger-ui`, `brick`, or `vty-windows`. I am not comfortable with pushing for windows availability of `hledger-ui` before fixing this bug.

**Comment by simonmichael** [comment]
Wonderful ! I was thinking about this, thanks for working on it.

**Comment by simonmichael** [comment]
(It will probably not get merged in master until stackage nightly gets brick 2.)

**Comment by simonmichael** [comment]
I'm happy to handle the CI part.

**Comment by simonmichael** [comment]
(Sorry for the stream of messages, last one:)

It's worth testing in as many windows terminal environments as you can find, we'll probably get different bug reports for each one.

**Comment by amano-kenji** [comment]
Did you try latest versions of vty and brick? The latest versions might have fixed the rendering issue.

**Comment by ShrykeWindgrace** [comment]
> (Sorry for the stream of messages, last one:)
> 
> It's worth testing in as many windows terminal environments as you can find, we'll probably get different bug reports for each one.

I guess most of this bugreports will be safely redirected to vty/vty-windows in that case, with all being "ok, in that terminal emulator it almost works, except for that and that feature" =) We can explicitly mention that `hledger-ui` will be ok for Windows Terminal, ok-ish in `ListOfTerminalEmulators`, and most probably not ok in `AnotherListOfTerminalEmulators`.

**Comment by ShrykeWindgrace** [comment]
> Did you try latest versions of vty and brick? The latest versions might have fixed the rendering issue.

Yes, I tested both master and latest releases. This is not a rendering issue, it is a "restore terminal to its prior state" issue.

**Comment by ShrykeWindgrace** [comment]
Well, for now the list of supported terminal emulators is a WIP, https://github.com/chhackett/vty-windows/wiki/TerminalSupport with latest info. The bug with terminal state is fixed, I'll open the PR in the next couple of days.

**Comment by the-solipsist** [comment]
I'm having trouble with hledger-ui / hledger-iadd and [kitty terminal](https://github.com/kovidgoyal/kitty/discussions/6718), and I think it might be because of brick/vty. Would anyone be in a position to confirm if this is the case?

**Comment by ShrykeWindgrace** [comment]
@the-solipsist in order to answer your question, we will need quite a list of things:

1. versions of hledger-ui/hledger-iadd you use

Then, if the answer to the first question is not "I use the custom build of these tools with vty-6 and brick-2", I suggest you open a separate issue in hledger.

Afterwards, this info is needed, too: 

2. version of `kitty` (most probably I won't be able to test that terminal emulator for quite some time)
3. your fonts
4. MWE for your workflow reproducing the bug
5. optionally, an attempt to reproduce the bug in _another_ terminal emulator (st/wezterm/foot/konsole/gnome_terminal/whatever you have access to). If the bug is no longer present, then the problem is most probably is in `kitty` or in your config of `kitty`.

**Comment by the-solipsist** [comment]
1. Hledger versions:
```
hledger --version && hledger-ui --version && hledger-iadd --version
hledger 1.32, linux-x86_64
hledger-ui 1.32, linux-x86_64
This is hledger-iadd version 1.3.19
```
These were installed from the Arch User Repository, using `hledger-bin`, `hledger-iadd-bin`, etc., which in turn use the Github builds of `hledger`, `hledger-ui`, `hledger-iadd`, etc.

2. Kitty version:
```
kitty --version
kitty 0.31.0 created by Kovid Goyal
```
3. Kitty fonts
```
kitty --debug-font-fallback
[341 10:55:57.150940] Preloaded font faces:
[341 10:55:57.150977] normal face: /usr/share/fonts/noto/NotoSansMono-Regular.ttf:0
[341 10:55:57.150998] bold face: /usr/share/fonts/noto/NotoSansMono-Bold.ttf:0
[341 10:55:57.151014] italic face: /usr/share/fonts/noto/NotoSansMono-Regular.ttf:0
[341 10:55:57.151031] bi face: /usr/share/fonts/noto/NotoSansMono-Bold.ttf:0
[341 10:55:57.305236] [glfw error 65544]: process_desktop_settings: failed with error: [org.freedesktop.DBus.Error.UnknownMethod] No such interface “org.freedesktop.portal.Settings” on object at path /org/freedesktop/portal/desktop
```
4. Reproducing the bug:
- Run `hledger bal` on a file with the `₹` symbol. The `₹` symbol shows fine.
- Open `hledger-ui`. Close `hledger-ui`.
- Run `hledger bal` again on the same file. The `₹` symbol is corrupted.

5. The bug isn't present in Gnome Console (which is why I thought the bug is with Kitty).

**Comment by ShrykeWindgrace** [comment]
@the-solipsist And what's the value  of `TERM` in both cases? 

@simonmichael could you move the thread on kitty to a separate issue? It has nothing to do with `vty-6` or `brick-2`. If there is no automatic option, I'll be able to simply copy all comments by hand, no trouble.

**Comment by simonmichael** [comment]
> could you move the thread on kitty to a separate issue?

Done, #2126.

Related to this issue, hledger-ui is now [out of stackage nightly](https://github.com/commercialhaskell/stackage/issues/7180#issuecomment-1845961656). I'm assuming supporting both old and new vty in the same release would be too much CPP and hassle (?), then getting it back in will require a new release that supports and requires vty 6.1+ and brick 2.1.1+.  Our 1.32.1 bugfix release is pending. But maybe squeezing a big vty/brick upgrade into a bugfix release is a bad idea.

**Comment by simonmichael** [comment]
I might do a separate hledger-ui-1.32.1 release after this has landed and we have had some cross platform user testing. (Need to figure out simpler release process for individual packages..)

**Comment by simonmichael** [comment]
Though, windows support is a feature worth shouting about, not a bugfix..

**Comment by ShrykeWindgrace** [comment]
Okies, we got `vty-6`, `vty-crossplatform` and `brick-2` landed into stackage, I'll prepare a PR with my changes

**Comment by amano-kenji** [comment]
I'm waiting for brick 2.3 support.
